### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by panianchuk**
https://panianchuk.github.io/react_todo-app-loading-todos/

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,223 +1,99 @@
 /* eslint-disable jsx-a11y/control-has-associated-label */
-import React, { useContext, useEffect, useRef } from 'react';
+// eslint-disable-next-line object-curly-newline
+import React, { useContext, useEffect, useRef, useState } from 'react';
+import { getTodos, createTodo } from './api/todos';
+import { Todo } from './types/Todo';
 import { AuthContext } from './components/Auth/AuthContext';
+import { Header } from './components/Header';
+import { TodoList } from './components/TodoList';
+import { Footer } from './components/Footer';
```
Comment (SanyaBratashchuk) [inline src/App.tsx:9]: Create index.ts file in components folder and export each component from it

File: src/App.tsx [hunk 2]
```diff
@@ -1,223 +1,99 @@
...
+import { ErrorNotification } from './components/ErrorNotification';
+import { filterTotos } from './api/filter';
+import { Filter } from './types/Filter';
+import { Error } from './types/Error';
```
Comment (SanyaBratashchuk) [inline src/App.tsx:13]: The same with types. Create index.ts file

File: src/App.tsx [hunk 3]
```diff
@@ -1,223 +1,99 @@
...
 
 export const App: React.FC = () => {
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const user = useContext(AuthContext);
```
Comment (SanyaBratashchuk) [inline src/App.tsx:17]: Move useContext(AuthContext) to custom useAuth hook

File: src/App.tsx [hunk 4]
```diff
@@ -1,223 +1,99 @@
...
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [title, setTitle] = useState('');
+  const [typeFilter, setTypeFilter] = useState(Filter.all);
+  const [error, setError] = useState('');
+  const [isHidden, setIsHidden] = useState(true);
```
Comment (SanyaBratashchuk) [inline src/App.tsx:22]: What is hidden? Todos? Filters? Specify that

File: src/App.tsx [hunk 5]
```diff
@@ -1,223 +1,99 @@
...
+          setError(Error.loading);
+          setIsHidden(false);
+          setTimeout(() => {
+            setIsHidden(true);
+          }, 3000);
+        });
+    }
   }, []);
 
+  const handleAddToto = (titleTodo: string) => {
```
Comment (SanyaBratashchuk) [inline src/App.tsx:45]:
```suggestion
  const handleAddTodo = (titleTodo: string) => {
```

File: src/api/filter.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import { Todo } from '../types/Todo';
+import { Filter } from '../types/Filter';
+
+const { active, completed } = Filter;
+
+export function filterTotos(todosArr: Todo[], type: string) {
+  switch (type) {
+    case completed:
```
Comment (SanyaBratashchuk) [inline src/api/filter.ts:8]:
```suggestion
    case Filter.active:
```

Comment (SanyaBratashchuk) [reply src/api/filter.ts:8]: Fix it everywhere

File: src/api/filter.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { Todo } from '../types/Todo';
+import { Filter } from '../types/Filter';
+
+const { active, completed } = Filter;
```
Comment (SanyaBratashchuk) [inline src/api/filter.ts:4]:
```suggestion
```

File: src/components/ErrorNotification.tsx [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+
+type Props = {
+  isHidden: boolean;
+  setIsHidden: (isHidden: boolean) => void;
```
Comment (SanyaBratashchuk) [inline src/components/ErrorNotification.tsx:5]:
```suggestion
  setIsHidden: Dispatch<SetStateAction<boolean>>;
```

**Review by SanyaBratashchuk — CHANGES_REQUESTED** [review]

File: src/components/Header.tsx [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+type Props = {
+  newTodoField: RefObject<HTMLInputElement>;
+  title: string;
+  setTitle: (title: string) => void;
+  handleAddToto: (title: string) => void;
+};
+
+/* eslint-disable jsx-a11y/control-has-associated-label */
+export const Header: React.FC<Props> = ({
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/components/Header.tsx:11]: if you pass ref - you have to wrap your component with `forwardRef()` and take it as second argument 
```suggestion
export const Header: React.FC<Props> = forwardRef((props, passedRef) => ...jsx)
```

File: src/App.tsx [hunk 10]
```diff
@@ -1,223 +1,103 @@
...
+  const [isHiddenErrorNote, setIsHiddenErrorNote] = useState(true);
+
   const newTodoField = useRef<HTMLInputElement>(null);
 
   useEffect(() => {
     // focus the element with `ref={newTodoField}`
     if (newTodoField.current) {
       newTodoField.current.focus();
     }
+
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/App.tsx:35]: Why just don't set `autofocus `attribute for input?

File: src/App.tsx [hunk 11]
```diff
@@ -1,223 +1,103 @@
...
+    };
+
+    createTodo(todo)
+      .then((loadedTodos) => setTodos(loadedTodos))
+      .catch(() => {
+        setError(Error.add);
+        setIsHiddenErrorNote(false);
+        setTimeout(() => {
+          setIsHiddenErrorNote(true);
+        }, 3000);
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/App.tsx:68]: Duplicated code from 41-44 lines, it is better to move these commands into separated function

File: src/App.tsx [hunk 12]
```diff
@@ -1,223 +1,103 @@
...
-      >
-        <button
-          data-cy="HideErrorButton"
-          type="button"
-          className="delete"
+        <Header
+          newTodoField={newTodoField}
+          title={title}
+          setTitle={setTitle}
+          handleAddToto={handleAddToto}
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/App.tsx:85]: props handler callback should be named as on...: onChange, onClick, onAddTodo.. etc
```suggestion
         onAddToto={handleAddToto}
```

File: src/components/TodoItem.tsx [hunk 13]
```diff
@@ -0,0 +1,46 @@
+import React from 'react';
+import classNames from 'classnames';
+import { Todo } from '../types';
+
+type Props = {
+  todo: Todo;
+};
+
+export const TodoItem: React.FC<Props> = ({ todo }) => {
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/components/TodoItem.tsx:9]:
```suggestion
export const TodoItem: React.FC<Props> = ({ todo: { completed , ...} }) => {
```

File: src/components/TodoList.tsx [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+    <section
+      className="todoapp__main"
+      data-cy="TodoList"
+    >
+      {todos.map((todo) => (
+        <TodoItem
+          todo={todo}
+          key={todo.id}
+        />
+      ))}
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/components/TodoList.tsx:20]:
```suggestion
      {todos.map((todo) => (
        <TodoItem
          key={todo.id}
          todo={todo}
        />
      ))}
```

File: src/types/Error.ts [hunk 15]
```diff
@@ -0,0 +1,4 @@
+export enum Error {
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/types/Error.ts:1]:
```suggestion
export enum ErrorTypes {
```

File: src/types/Error.ts [hunk 16]
```diff
@@ -0,0 +1,4 @@
...
+  loading = 'Unable to load a todos',
+  add = 'Unable to add a todo',
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/types/Error.ts:3]: first letter of enum keys has to be uppercase
```suggestion
  Loading = 'Unable to load a todos',
  Add = 'Unable to add a todo',
```

File: src/types/Filter.ts [hunk 17]
```diff
@@ -0,0 +1,5 @@
+export enum Filter {
+  all = 'all',
+  active = 'active',
+  completed = 'completed',
+}
```
Comment (Volodymyr-Melnyk-Fron10der) [inline src/types/Filter.ts:5]: First letter - uppercase

**Review by Volodymyr-Melnyk-Fron10der — CHANGES_REQUESTED** [review]
Hi, there is more comments from me)

**Review by qoa11a — APPROVED** [review]

**Pull request closed without merging**